### ISSUE TITLE
Can't encode MAX_INT64

### ISSUE THREAD

**Issue by pldubouilh** [feross/bittorrent-dht#115]
According to the BEP44, we should be able to set the token to MAX_INT64, (i.e. 0x7fffffffffffffff) when setting mutable content on the DHT. Although, it's not possible at the moment, because of an [issue in node-bencode](https://github.com/themasch/node-bencode/issues/35). Well actually it's more of a node issue, as it can't properly do very large numbers...  

OK, it's a bit of a corner case, but I'd like to see it working ! And [libtorrent treat it correctly](https://github.com/arvidn/libtorrent/blob/master/test/test_bencoding.cpp#L242).

**Comment by mafintosh** [comment]
which token? the write tokens are a buffer if i recall correctly

**Comment by pldubouilh** [comment]
Yep, sorry token's probably not the right terminology here. I meant the sequence number, [like here](https://github.com/pldubouilh/chunks-server/blob/master/postLatest.js#L140).

**Comment by mafintosh** [comment]
ah i see :) thats probably more a theoretical problem than an actual one. i mean assuming you do one update per ms Date.now() will overflow before this becomes a problem hehe

**Comment by mafintosh** [comment]
64bit ints in js would be great though (!)

**Comment by pldubouilh** [comment]
Yes, the issue's theoretical, and there's unfortunately no easy fix ! I tried to come up with a ```solution''' around node-bencode, but it's not pretty - at all !

The thing is nothing's preventing you to make higher gaps. I was actually aiming at "locking" a given dht spot by setting the token to MAX_INT64. And it seems that other DHT implementations are taking it fine. 

I'm also concerned about treating such a request. I'm afraid a request emitted for instance by a libtorrent node, containing that MAX_INT64 sequence number, would be truncated by bittorent-dht when decoded... If anyone (apart from me) is actually crazy enough to use that weird option :)

**Comment by mafintosh** [comment]
assuming we can add a test for this somehow we should make that bittorrent-dht at least doesn't blow up. @feross any ideas on simple ways we could test this?

**Comment by pldubouilh** [comment]
Hmm, I don't see any easy way to test that. Trying to dht.set MAX_INT64 will throw [this](https://github.com/themasch/node-bencode/blob/master/lib/encode.js#L56) in node-bencode. 

And dht.get, well, I guess it'll be interpreted as Number.MAX_SAFE_INTEGER, which is fine as it won't blow up, but not nice as it's not what's on the DHT. But again, weird edge case !

**Comment by feross** [comment]
Sounds like @pldubouilh has just described a test case that we'd current fail on:

> Trying to dht.set MAX_INT64 will throw this in node-bencode.

**Comment by mafintosh** [comment]
@feross how? we try-catch incoming bencoded messages iirc

**Comment by feross** [comment]
@mafintosh Ahh cool, then we're probably fine just not worrying about this issue. Or, we could try using https://github.com/indutny/bn.js/ but that might be overkill for this edge case.

**Comment by pldubouilh** [comment]
I had a go at making node-bencode work with big-numbers, and it's doing the job, although it's not the most elegant solution... It's a bit slower as well when encoding but that will probably not be an issue here.

But on the other hand the only moment where integers are bencoded with bittorrent-dht are with mutable puts, so the annoyance of using node-bencode with big-numbers would only be limited to that operation. And that would give bittorrent-dht full compliance to bep44 :)

**Comment by feross** [comment]
@pldubouilh Any progress getting `bencode` to work with big numbers? It looks like this lack of support is probably causing another, worse issue: https://github.com/DiegoRBaquero/BTorrent/issues/26

Would be great to get it fixed up!

**Comment by pldubouilh** [comment]
Got a pretty busy couple of weeks, so it should be coming this week !

**Comment by pldubouilh** [comment]
The issue here is this  :
``` javascript
encode.number = function( buffers, data ) {
  var maxLo = 0x80000000
  var hi = ( data / maxLo ) << 0
  var lo = ( data % maxLo  ) << 0
  var val = hi * maxLo + lo

  buffers.push( new Buffer( 'i' + val + 'e' ))
  [...]
}
```
For some reason (that'd actually be interesting to investigate) `val` is computed incorrectly on ARM, and resulting with an incorrecect bencoded value. To fix it @solderzzc directly put `buffers.push( new Buffer( 'i' + data + 'e' ))`, bypassing the overflow detection thing. 

Using my bignumber-friendly bencode, would likely fix that as I basically removed that overflow detection bit. So it's the issue is not directly related, but the fix would be the same.

RE the impact to the API, that will be limited to mutable put as it's the only time where numbers are bencoded by bittorrent-dht. Basically instead of having a number in opts.seq, that will be a BigNumber. 

My bignumber+bencode is [here](https://github.com/pldubouilh/node-bencode/commit/c84e3b0bb17a874625aa5c940d1aba5479ce123c) and I have it running with bittorrent-dht without any issue. I'll try running both versions on a raspberry pi see if I can reproduce that bug.

**Comment by mafintosh** [comment]
sounds like a bencode (or even arm bug) that is unrelated to this? i'm a bit critical about adding big num in general if it's only to solve a theoretical use case with bep 44

**Comment by feross** [comment]
Yeah, agree. It's too much overheard to force the user to use a bignum library. I was just hoping that this could be solved internally to `bencode` so it doesn't do the wrong thing on ARM.

**Comment by nazar-pc** [comment]
https://github.com/themasch/node-bencode/issues/35 is fixed, should this be closed now?
